pkg/scheduler/framework/plugins/dynamicresources/structuredparameters.go

```diff
@@ -0,0 +1,235 @@
+	// multiple requests which need to be allocated together, so here
+	// we have to collect them per model.
+	type perDriverRequests struct {
+		parameters []runtime.RawExtension
+		requests   []*resourcev1alpha2.NamedResourcesRequest
+	}
+	namedresourcesRequests := make(map[string]perDriverRequests)
+	for i, request := range claimParameters.DriverRequests {
+		driverName := request.DriverName
+		p := namedresourcesRequests[driverName]
+		for e, request := range request.Requests {
+			switch {
+			case request.ResourceRequestModel.NamedResources != nil:
+				p.parameters = append(p.parameters, request.VendorParameters)
+				p.requests = append(p.requests, request.ResourceRequestModel.NamedResources)
+			default:
+				return nil, fmt.Errorf("claim parameters %s: driverRequersts[%d].requests[%d]: no supported structured parameters found", klog.KObj(claimParameters), i, e)
+			}
+		}
+		if len(p.requests) > 0 {
+			namedresourcesRequests[driverName] = p
+		}
+	}
+
+	c := &claimController{
+		class:           class,
+		classParameters: classParameters,
+		claimParameters: claimParameters,
+		namedresources:  make(map[string]perDriverController, len(namedresourcesRequests)),
+	}
+	for driverName, perDriver := range namedresourcesRequests {
+		var filter *resourcev1alpha2.NamedResourcesFilter
+		if classParameters != nil {
+			for _, f := range classParameters.Filters {
+				if f.DriverName == driverName && f.ResourceFilterModel.NamedResources != nil {
+					filter = f.ResourceFilterModel.NamedResources
+					break
+				}
+			}
+		}
+		controller, err := namedresourcesmodel.NewClaimController(filter, perDriver.requests)
+		if err != nil {
+			return nil, fmt.Errorf("creating claim controller for named resources structured model: %w", err)
+		}
+		c.namedresources[driverName] = perDriverController{
+			parameters: perDriver.parameters,
+			controller: controller,
+		}
+	}
+	return c, nil
+}
+
+// claimController currently wraps exactly one structured parameter model.
+
+type claimController struct {
+	class           *resourcev1alpha2.ResourceClass
+	classParameters *resourcev1alpha2.ResourceClassParameters
+	claimParameters *resourcev1alpha2.ResourceClaimParameters
+	namedresources  map[string]perDriverController
+}
+
+type perDriverController struct {
+	parameters []runtime.RawExtension
+	controller *namedresourcesmodel.Controller
+}
+
+func (c claimController) nodeIsSuitable(ctx context.Context, nodeName string, resources resources) (bool, error) {
+	nodeResources := resources[nodeName]
+	for driverName, perDriver := range c.namedresources {
+		okay, err := perDriver.controller.NodeIsSuitable(ctx, nodeResources[driverName].namedresources)
+		if err != nil {
+			// This is an error in the CEL expression which needs
+			// to be fixed. Better fail very visibly instead of
+			// ignoring the node.
+			return false, fmt.Errorf("checking node %q and resources of driver %q: %w", nodeName, driverName, err)
```

Comment:
Would this user error be avoided by API validation?

If not, this should lead to UnschedulableAndUnresolvable, not Error.

Reply:
CEL expressions can reference attributes without checking that they exist. This then leads to a map lookup runtime error, depending on the actual resource attributes. Such errors cannot be found by validation because validation has no knowledge about vendor attributes and which of those are optional and which are required.

But because typically vendors are producing these CEL expressions, they should never fail at runtime. I'm still leaning towards returning an error here and dealing with it as an error status in `dynamicResources.Filter`.

Reply:
But if someone misuses the feature directly, it could lead to unnecessary scheduling retries.

Reply:
That's a good point. Let me change how the error is handled in `dynamicResources.Filter`.